lib/dry/schema/extensions/json_schema/schema_compiler.rb

```diff
@@ -0,0 +1,182 @@
+# frozen_string_literal: true
+
+require "dry/schema/constants"
+
+module Dry
+  module Schema
+    # @api private
+    module JSONSchema
+      # @api private
+      class SchemaCompiler
+        IDENTITY = ->(v) { v }
+
+        PREDICATE_TO_TYPE = {
+          array?: {type: "array"},
+          bool?: {type: "boolean"},
+          date?: {type: "string", format: "date"},
+          date_time?: {type: "string", format: "date-time"},
+          decimal?: {type: "number"},
+          float?: {type: "number"},
+          hash?: {type: "object"},
+          int?: {type: "integer"},
+          nil?: {type: "null"},
+          str?: {type: "string"},
+          time?: {type: "string", format: "time"},
+          min_size?: {minLength: ->(v) { v.to_i }},
+          max_size?: {maxLength: ->(v) { v.to_i }},
+          included_in?: {enum: ->(v) { v.to_a }},
+          uri?: {format: "uri"},
+          uuid_v1?: {
+            pattern: "^[0-9A-F]{8}-[0-9A-F]{4}-1[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$"
+          },
+          uuid_v2?: {
+            pattern: "^[0-9A-F]{8}-[0-9A-F]{4}-2[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$"
+          },
+          uuid_v3?: {
+            pattern: "^[0-9A-F]{8}-[0-9A-F]{4}-3[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$"
+          },
+          uuid_v4?: {
+            pattern: "^[a-f0-9]{8}-?[a-f0-9]{4}-?4[a-f0-9]{3}-?[89ab][a-f0-9]{3}-?[a-f0-9]{12}$"
+          },
+          uuid_v5?: {
+            pattern: "^[0-9A-F]{8}-[0-9A-F]{4}-5[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$"
+          },
+          gt?: {exclusiveMinimum: IDENTITY},
+          gteq?: {mininum: IDENTITY},
+          lt?: {exclusiveMaximum: IDENTITY},
+          lteq?: {maximum: IDENTITY},
+          odd?: {type: "integer", not: {multipleOf: 2}},
+          even?: {type: "integer", multipleOf: 2}
+        }.freeze
+
+        # @api private
+        attr_reader :keys, :required
+
+        # @api private
+        def initialize(root: false)
+          @keys = EMPTY_HASH.dup
+          @required = Set.new
+          @root = root
+        end
+
+        # @api private
+        def to_h
+          result = {}
+          result[:$schema] = "http://json-schema.org/draft-06/schema#" if root?
+          result.merge!(type: "object", properties: keys, required: required.to_a)
+          result
+        end
+
+        # @api private
+        def call(ast)
+          visit(ast)
+        end
+
+        # @api private
+        def visit(node, opts = EMPTY_HASH)
+          meth, rest = node
+          public_send(:"visit_#{meth}", rest, opts)
+        end
+
+        # @api private
+        def visit_set(node, opts = EMPTY_HASH)
+          target = (key = opts[:key]) ? self.class.new : self
```

Comment:
I a bunch of places we're using `opts[]` instead of `opts.key?`. Do we have legit usecases where opts could be set and be `nil`/`false` at the same time?
IMHO would be more clear to use `#key?`.